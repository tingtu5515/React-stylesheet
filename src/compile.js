/**
 * @flow
 */

import type {StylesheetSpec} from './Stylesheet';

import prefix from 'inline-style-prefix-all';
import createHash from 'murmurhash-js/murmurhash3_gc';
import hyphenateStyleName from 'fbjs/lib/hyphenateStyleName';
import memoizeStringOnly from 'fbjs/lib/memoizeStringOnly';

export type ClassNameMapping = {
  className: string;
  then?: {[name: string]: ClassNameMapping};
};

export type CompileResult = {
  id: string;
  css: string;
  mapping: ClassNameMapping;
};

const UNITLESS_NUMBER = {
  animationIterationCount: true,
  borderImageOutset: true,
  borderImageSlice: true,
  borderImageWidth: true,
  boxFlex: true,
  boxFlexGroup: true,
  boxOrdinalGroup: true,
  columnCount: true,
  flex: true,
  flexGrow: true,
  flexPositive: true,
  flexShrink: true,
  flexNegative: true,
  flexOrder: true,
  gridRow: true,
  gridColumn: true,
  fontWeight: true,
  lineClamp: true,
  lineHeight: true,
  opacity: true,
  order: true,
  orphans: true,
  tabSize: true,
  widows: true,
  zIndex: true,
  zoom: true,

  // SVG-related properties
  fillOpacity: true,
  floodOpacity: true,
  stopOpacity: true,
  strokeDasharray: true,
  strokeDashoffset: true,
  strokeMiterlimit: true,
  strokeOpacity: true,
  strokeWidth: true
};

export const PSEUDO_CLASS = {
  focus: true,
  hover: true,
  active: true,
  checked: true,
  default: true,
  disabled: true,
  empty: true,
  enabled: true,
  firstChild: true,
  firstOfType: true,
  fullscreen: true,
  indeterminate: true,
  invalid: true,
  lastChild: true,
  lastOfType: true,
  link: true,
  onlyChild: true,
  optional: true,
  required: true,
  root: true,
  scope: true,
  target: true,
  valid: true,
  visited: true,
};

export default function compile(
  name: string,
  stylesheet: StylesheetSpec
): CompileResult {

  let id = createHash(JSON.stringify(stylesheet));
  let mapping = {};
  let css = [];

  // If base is not defined we need to define it so defaults style kicks in.
  if (stylesheet.base == null) {
    stylesheet = {
      base: {},
      ...stylesheet,
    };
  }

  for (let variant in stylesheet) {

    let style = stylesheet[variant];
    if (variant.indexOf('_') > -1) {
      variant = variant.split('_');
    } else {
      variant = [variant];
    }
    variant.sort();

    let seenPseudoClass = false;
    let trace = [name];
    let variantMapping = mapping;
    for (let i = 0; i < variant.length; i++) {
      if (variant[i] === 'base' && variant.length > 1) {
        throw new Error('"base" should not be used as a part of a complex variant');
      }

      if (variant[i] !== 'base' && !seenPseudoClass) {
        variantMapping.then = variantMapping.then || {};
        variantMapping.then[variant[i]] = variantMapping.then[variant[i]] || {};
        variantMapping = variantMapping.then[variant[i]];
        trace.push(variant[i]);
      }

      if (i === variant.length - 1) {
        if (variant[i] === 'base') {
          style = {boxSizing: 'border-box', ...style};
        }
        let className = `${trace.join('-')}-${id}`;
        css.push(compileStyle(className, style));
        variantMapping.className = className;
      }
    }

  }

  return {id, css: css.join('\n'), mapping};
}

function compileStyle(className, style) {
  let css = [];
  let cssList = [];
  let ownStyle = {};

  for (let name in style) {
    let value = style[name];
    if (!style.hasOwnProperty(name)) {
      continue;
    }
    if (PSEUDO_CLASS.hasOwnProperty(name) && PSEUDO_CLASS[name]) {
      // this is pseudo class, recurse
      cssList.push(compileStyle(className + ':' + compileName(name), value));
    } else if (!isEmpty(value)) {
      ownStyle[name] = value;
    }
  }

  ownStyle = prefix(ownStyle);

  for (let name in ownStyle) {
    let value = ownStyle[name];
    if (Array.isArray(value)) {
      for (let i = 0; i < value.length; i++) {
        css.push(...compileItem(name, value[i]));
      }
    } else {
      css.push(...compileItem(name, value));
    }
  }

  if (css.length > 0) {
    return [`.${className} { ${css.join(';')}; }`].concat(cssList).join('\n');
  } else {
    return cssList.join('\n');
  }
}

function isEmpty(value) {
  return (
    value == null ||
    value === '' ||
    value === false
  );
}

/**
 * Compile item.
 */
function compileItem(name, value) {
  switch (name) {
    case 'paddingH':
      return [
        compileItem('paddingLeft', value),
        compileItem('paddingRight', value)
      ];
    case 'paddingV':
      return [
        compileItem('paddingTop', value),
        compileItem('paddingBottom', value)
      ];
    case 'marginH':
      return [
        compileItem('marginLeft', value),
        compileItem('marginRight', value)
      ];
    case 'marginV':
      return [
        compileItem('marginTop', value),
        compileItem('marginBottom', value)
      ];
    default: {
      let cssName = compileName(name);
      let cssValue = compileValue(name, value);
      return [`${cssName}:${cssValue}`];
    }
  }
}

/**
 * Compile style prop name.
 *
 * Based on code in React, see react/lib/CSSPropertyOperations module.
 */
let compileName = memoizeStringOnly(name => hyphenateStyleName(name));

/**
 * Compile style prop value.
 *
 * Based on code in React, see react/lib/dangerousStyleValue module.
 */
function compileValue(name: string, value: mixed): string {
  let isNonNumeric = isNaN(value);
  if (
    isNonNumeric ||
    value === 0 ||
    UNITLESS_NUMBER.hasOwnProperty(name) && UNITLESS_NUMBER[name]
  ) {
    return '' + ((value: any): string); // cast to string
  } else {
    return ((value: any): string) + 'px';
  }
}
