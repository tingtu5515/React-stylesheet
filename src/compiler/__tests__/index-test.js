import compile from '../index';

it('compiles base', function() {
  let stylesheet = {
    base: {
      color: 'red'
    }
  };
  let {id, css, mapping} = compile('name', stylesheet);
  expect(mapping).toEqual({
    className: `name-${id}`
  });
  expect(css).toMatchSnapshot();
});

it('compiles to hyphenated name', function() {
  let stylesheet = {
    base: {
      textOverflow: 'ellipsis'
    }
  };
  let {css} = compile('name', stylesheet);
  expect(css).toMatchSnapshot();
});

it('compiles to numbers to px', function() {
  let stylesheet = {
    base: {
      width: 10
    }
  };
  let {css} = compile('name', stylesheet);
  expect(css).toMatchSnapshot();
});

it('keeps numbers for unitless properties', function() {
  let stylesheet = {
    base: {
      order: 10
    }
  };
  let {css} = compile('name', stylesheet);
  expect(css).toMatchSnapshot();
});

it('skips empty values', function() {
  let stylesheet = {
    base: {
      color: 'red',
      background: null,
      textAlign: false,
      flex: '',
      flexDirection: undefined,
    }
  };
  let {css} = compile('name', stylesheet);
  expect(css).toMatchSnapshot();
});

it('handles arrays', function() {
  let stylesheet = {
    base: {
      color: ['red', 'white'],
    }
  };
  let {css} = compile('name', stylesheet);
  expect(css).toMatchSnapshot();
});

it('compiles base w/ variant', function() {
  let stylesheet = {
    base: {
      color: 'red'
    },
    em: {
      color: 'green'
    }
  };
  let {id, css, mapping} = compile('name', stylesheet);
  expect(mapping).toEqual({
    className: `name-${id}`,
    then: {
      em: {
        className: 'name-em-2845490580'
      }
    }
  });
  expect(css).toMatchSnapshot();
});

it('compiles variant', function() {
  let stylesheet = {
    em: {
      color: 'red'
    }
  };
  let {id, css, mapping} = compile('name', stylesheet);
  expect(mapping).toEqual({
    className: `name-${id}`,
    then: {
      em: {
        className: `name-em-${id}`
      }
    }
  });
  expect(css).toMatchSnapshot();
});

it('compiles base w/ pseudo', function() {
  let stylesheet = {
    base: {
      color: 'red',
      hover: {
        color: 'white'
      }
    },
  };
  let {id, css, mapping} = compile('name', stylesheet);
  expect(mapping).toEqual({
    className: `name-${id}`
  });
  expect(css).toMatchSnapshot();
});

it('compiles base w/ hyphenated pseudo', function() {
  let stylesheet = {
    base: {
      color: 'red',
      firstOfType: {
        color: 'white'
      }
    },
  };
  let {id, css, mapping} = compile('name', stylesheet);
  expect(mapping).toEqual({
    className: `name-${id}`
  });
  expect(css).toMatchSnapshot();
});

it('compiles pseudo', function() {
  let stylesheet = {
    base: {
      hover: {
        color: 'white'
      }
    }
  };
  let {id, css, mapping} = compile('name', stylesheet);
  expect(mapping).toEqual({
    className: `name-${id}`
  });
  expect(css).toMatchSnapshot();
});


it('compiles variant pseudo', function() {
  let stylesheet = {
    em: {
      hover: {
        color: 'white'
      }
    }
  };
  let {id, css, mapping} = compile('name', stylesheet);
  expect(mapping).toEqual({
    className: `name-${id}`,
    then: {
      em: {
        className: `name-em-${id}`
      }
    }
  });
  expect(css).toMatchSnapshot();
});

it('compiles variant w/ variant pseudo', function() {
  let stylesheet = {
    em: {
      color: 'red',
      hover: {
        color: 'white'
      }
    },
  };
  let {id, css, mapping} = compile('name', stylesheet);
  expect(mapping).toEqual({
    className: `name-${id}`,
    then: {
      em: {
        className: `name-em-${id}`
      }
    }
  });
  expect(css).toMatchSnapshot();
});

it('compiles double variant', function() {
  let stylesheet = {
    em_it: {
      color: 'white'
    }
  };
  let {id, css, mapping} = compile('name', stylesheet);
  expect(mapping).toEqual({
    className: `name-${id}`,
    then: {
      em: {
        then: {
          it: {
            className: `name-em-it-${id}`
          }
        }
      }
    }
  });
  expect(css).toMatchSnapshot();
});


it('compiles double variant w/ pseudo', function() {
  let stylesheet = {
    em_it: {
      hover: {
        color: 'white'
      }
    }
  };
  let {id, css, mapping} = compile('name', stylesheet);
  expect(mapping).toEqual({
    className: `name-${id}`,
    then: {
      em: {
        then: {
          it: {
            className: `name-em-it-${id}`
          }
        }
      }
    }
  });
  expect(css).toMatchSnapshot();
});

it('compiles double variant w/ separate variants', function() {
  let stylesheet = {
    it: {
      color: 'green',
    },
    em: {
      color: 'red',
    },
    em_it: {
      color: 'white'
    }
  };
  let {id, css, mapping} = compile('name', stylesheet);
  expect(mapping).toEqual({
    className: `name-${id}`,
    then: {
      em: {
        className: `name-em-${id}`,
        then: {
          it: {
            className: `name-em-it-${id}`
          }
        }
      },
      it: {
        className: `name-it-${id}`,
      }
    }
  });
  expect(css).toMatchSnapshot();
});

it('compiles double pseudo', function() {
  let stylesheet = {
    base: {
      hover: {
        focus: {
          color: 'red'
        }
      }
    },
  };
  let {id, css, mapping} = compile('name', stylesheet);
  expect(mapping).toEqual({
    className: `name-${id}`,
  });
  expect(css).toMatchSnapshot();
});

it('compiles double pseudo (intermediate)', function() {
  let stylesheet = {
    base: {
      color: 'black',
      hover: {
        color: 'white',
        focus: {
          color: 'red'
        }
      }
    },
  };
  let {id, css, mapping} = compile('name', stylesheet);
  expect(mapping).toEqual({
    className: `name-${id}`,
  });
  expect(css).toMatchSnapshot();
});

it('compiles variant w/ double pseudo (intermediate)', function() {
  let stylesheet = {
    em: {
      color: 'black',
      hover: {
        color: 'white',
        focus: {
          color: 'red'
        }
      }
    },
  };
  let {id, css, mapping} = compile('name', stylesheet);
  expect(mapping).toEqual({
    className: `name-${id}`,
    then: {
      em: {
        className: `name-em-${id}`,
      }
    }
  });
  expect(css).toMatchSnapshot();
});

it('compiles paddingH', function() {
  let stylesheet = {
    base: {
      paddingH: 10
    },
  };
  let {css} = compile('name', stylesheet);
  expect(css).toMatchSnapshot();
});

it('compiles paddingV', function() {
  let stylesheet = {
    base: {
      paddingV: 10
    },
  };
  let {css} = compile('name', stylesheet);
  expect(css).toMatchSnapshot();
});

it('compiles marginH', function() {
  let stylesheet = {
    base: {
      marginH: 10
    },
  };
  let {css} = compile('name', stylesheet);
  expect(css).toMatchSnapshot();
});

it('compiles marginV', function() {
  let stylesheet = {
    base: {
      marginV: 10
    },
  };
  let {css} = compile('name', stylesheet);
  expect(css).toMatchSnapshot();
});
