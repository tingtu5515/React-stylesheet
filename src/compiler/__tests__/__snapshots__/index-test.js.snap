// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`compiles base 1`] = `".name-917847219 { box-sizing:border-box;color:red;-moz-box-sizing:border-box; }"`;

exports[`compiles base w/ hyphenated pseudo 1`] = `
".name-1566645038 { box-sizing:border-box;color:red;-moz-box-sizing:border-box; }
.name-1566645038:first-of-type { color:white; }"
`;

exports[`compiles base w/ pseudo 1`] = `
".name-2233461878 { box-sizing:border-box;color:red;-moz-box-sizing:border-box; }
.name-2233461878:hover { color:white; }"
`;

exports[`compiles base w/ variant 1`] = `
".name-2845490580 { box-sizing:border-box;color:red;-moz-box-sizing:border-box; }
.name-em-2845490580 { color:green; }"
`;

exports[`compiles border: {width: 1, color: "#aaa"} 1`] = `".name-1877212110 { box-sizing:border-box;border:1px none #aaa;-moz-box-sizing:border-box; }"`;

exports[`compiles borderLeft: {width: 1, color: "#aaa"} 1`] = `".name-3941833314 { box-sizing:border-box;border-left:1px none #aaa;-moz-box-sizing:border-box; }"`;

exports[`compiles boxShadow: [...] 1`] = `".name-1541599391 { box-sizing:border-box;box-shadow:inset 10px 10px 2px 0px #000, 10px 10px 2px 0px #000;-moz-box-sizing:border-box; }"`;

exports[`compiles boxShadow: [string, syntax] 1`] = `".name-1307827675 { box-sizing:border-box;box-shadow:inset 10px 10px 2px 0px #000000, 10px 10px 2px 0px #000;-moz-box-sizing:border-box; }"`;

exports[`compiles boxShadow: {inset: true, x: 10, y: 10, blur: 2} 1`] = `".name-852133784 { box-sizing:border-box;box-shadow:inset 10px 10px 2px 0px #000;-moz-box-sizing:border-box; }"`;

exports[`compiles boxShadow: {x: 10, y: 10, blur: 2} 1`] = `".name-3203728316 { box-sizing:border-box;box-shadow:10px 10px 2px 0px #000;-moz-box-sizing:border-box; }"`;

exports[`compiles double pseudo (intermediate) 1`] = `
".name-2319864360 { box-sizing:border-box;color:black;-moz-box-sizing:border-box; }
.name-2319864360:hover { color:white; }
.name-2319864360:hover:focus { color:red; }"
`;

exports[`compiles double pseudo 1`] = `
".name-412780474 { box-sizing:border-box;-moz-box-sizing:border-box; }
.name-412780474:hover:focus { color:red; }"
`;

exports[`compiles double variant 1`] = `
".name-88266418 { box-sizing:border-box;-moz-box-sizing:border-box; }
.name-em-it-88266418 { color:white; }"
`;

exports[`compiles double variant w/ pseudo 1`] = `
".name-1352891476 { box-sizing:border-box;-moz-box-sizing:border-box; }
.name-em-it-1352891476:hover { color:white; }"
`;

exports[`compiles double variant w/ separate variants 1`] = `
".name-3975713384 { box-sizing:border-box;-moz-box-sizing:border-box; }
.name-it-3975713384 { color:green; }
.name-em-3975713384 { color:red; }
.name-em-it-3975713384 { color:white; }"
`;

exports[`compiles margin: {top: 10, left: 20} 1`] = `".name-3870247771 { box-sizing:border-box;margin:10px 0  0 20px;-moz-box-sizing:border-box; }"`;

exports[`compiles padding: {horizontal: 10} 1`] = `".name-4271658450 { box-sizing:border-box;padding:0 10px  0 10px;-moz-box-sizing:border-box; }"`;

exports[`compiles padding: {top: 10, left: 20} 1`] = `".name-1841606720 { box-sizing:border-box;padding:10px 0  0 20px;-moz-box-sizing:border-box; }"`;

exports[`compiles padding: {vertical: 10, top: 5} 1`] = `".name-3888212147 { box-sizing:border-box;padding:5px 0  10px 0;-moz-box-sizing:border-box; }"`;

exports[`compiles padding: {vertical: 10} 1`] = `".name-2758546622 { box-sizing:border-box;padding:10px 0  10px 0;-moz-box-sizing:border-box; }"`;

exports[`compiles pseudo 1`] = `
".name-2156999832 { box-sizing:border-box;-moz-box-sizing:border-box; }
.name-2156999832:hover { color:white; }"
`;

exports[`compiles textShadow: {x: 10, y: 10} 1`] = `".name-2860912185 { box-sizing:border-box;text-shadow:10px 10px 0px #000;-moz-box-sizing:border-box; }"`;

exports[`compiles to hyphenated name 1`] = `".name-2547837783 { box-sizing:border-box;text-overflow:ellipsis;-moz-box-sizing:border-box; }"`;

exports[`compiles to numbers to px 1`] = `".name-2380892734 { box-sizing:border-box;width:10px;-moz-box-sizing:border-box; }"`;

exports[`compiles transition: [...] 1`] = `".name-3026480385 { box-sizing:border-box;transition:margin 10s ease 0s, top 2s ease 0s;-moz-box-sizing:border-box;-webkit-transition:margin 10s ease 0s, top 2s ease 0s; }"`;

exports[`compiles transition: {property: "margin", duration: 10} 1`] = `".name-2218216753 { box-sizing:border-box;transition:margin 10s ease 0s;-moz-box-sizing:border-box;-webkit-transition:margin 10s ease 0s; }"`;

exports[`compiles variant 1`] = `
".name-2439838490 { box-sizing:border-box;-moz-box-sizing:border-box; }
.name-em-2439838490 { color:red; }"
`;

exports[`compiles variant pseudo 1`] = `
".name-2170411812 { box-sizing:border-box;-moz-box-sizing:border-box; }
.name-em-2170411812:hover { color:white; }"
`;

exports[`compiles variant w/ double pseudo (intermediate) 1`] = `
".name-1587164824 { box-sizing:border-box;-moz-box-sizing:border-box; }
.name-em-1587164824 { color:black; }
.name-em-1587164824:hover { color:white; }
.name-em-1587164824:hover:focus { color:red; }"
`;

exports[`compiles variant w/ variant pseudo 1`] = `
".name-887113354 { box-sizing:border-box;-moz-box-sizing:border-box; }
.name-em-887113354 { color:red; }
.name-em-887113354:hover { color:white; }"
`;

exports[`handles arrays 1`] = `".name-3158404022 { box-sizing:border-box;color:red;color:white;-moz-box-sizing:border-box; }"`;

exports[`keeps numbers for unitless properties 1`] = `".name-1768927200 { box-sizing:border-box;order:10;-moz-box-sizing:border-box;-webkit-order:10;-ms-flex-order:10; }"`;

exports[`skips empty values 1`] = `".name-3779141731 { box-sizing:border-box;color:red;-moz-box-sizing:border-box; }"`;
